Sample/HotelManagement/HotelManagement/GuestStayAccounts/GuestStayAccount.cs

```diff
@@ -0,0 +1,111 @@
+using HotelManagement.Core;
+using HotelManagement.Core.Structures;
+using static HotelManagement.Core.Structures.Result;
+
+namespace HotelManagement.GuestStayAccounts;
+
+public record GuestCheckedIn(
+    Guid GuestStayId,
+    DateTimeOffset CheckedInAt
+);
+
+public record ChargeRecorded(
+    Guid GuestStayId,
+    Decimal Amount,
+    DateTimeOffset RecordedAt
+);
+
+public record PaymentRecorded(
+    Guid GuestStayId,
+    Decimal Amount,
+    DateTimeOffset RecordedAt
+);
+
+public record GuestCheckedOut(
+    Guid GuestStayId,
+    DateTimeOffset CheckedOutAt,
+    Guid? GroupCheckOutId = null
+);
+
+public record GuestCheckoutFailed(
+    Guid GuestStayId,
+    GuestCheckoutFailed.FailureReason Reason,
+    DateTimeOffset FailedAt,
+    Guid? GroupCheckOutId = null
+)
+{
+    public enum FailureReason
+    {
+        BalanceNotSettled
+    }
+}
+
+public record GuestStayAccount(
+    Guid Id,
+    decimal Balance = 0,
+    GuestStayAccountStatus Status = GuestStayAccountStatus.Opened
+)
+{
+    public bool IsSettled => Balance == 0;
+
+
+    public static GuestCheckedIn CheckIn(Guid guestStayId, DateTimeOffset now) =>
+        new GuestCheckedIn(guestStayId, now);
+
+    public ChargeRecorded RecordCharge(decimal amount, DateTimeOffset now)
+    {
+        if (Status != GuestStayAccountStatus.Opened)
+            throw new InvalidOperationException("Cannot record charge for not opened account");
+
+        return new ChargeRecorded(Id, amount, now);
+    }
+
+    public PaymentRecorded RecordPayment(decimal amount, DateTimeOffset now)
+    {
+        if (Status != GuestStayAccountStatus.Opened)
+            throw new InvalidOperationException("Cannot record charge for not opened account");
+
+        return new PaymentRecorded(Id, amount, now);
+    }
+
+    public Result<GuestCheckedOut, GuestCheckoutFailed> CheckOut(DateTimeOffset now, Guid? groupCheckoutId = null)
+    {
+        if (Status != GuestStayAccountStatus.Opened)
+            throw new InvalidOperationException("Cannot record payment for not opened account");
+
+        return IsSettled
+            ? Success<GuestCheckedOut, GuestCheckoutFailed>(
+                new GuestCheckedOut(
+                    Id,
+                    DateTimeOffset.UtcNow,
```

Comment:
in other places it's common one, not UTC. Just wondering

Reply:
Should be UTC in all those places, as it's a timestamp of what has happened in the background. Fixed in https://github.com/oskardudycz/EventSourcing.NetCore/pull/167/commits/e93f1412470be074658a3324656bee4e92c0a952. You can read more about my perspective in https://event-driven.io/en/is_keeping_utc_dates_best_solution/.